packages/loader/container-loader/src/container.ts

```diff
@@ -244,30 +245,35 @@ export class CollabWindowTracker {
         }
 
         // We don't acknowledge no-ops to avoid acknowledgement cycles (i.e. ack the MSN
-        // update, which updates the MSN, then ack the update, etc...).
-        if (message.type === MessageType.NoOp) {
+        // update, which updates the MSN, then ack the update, etc...). Also, don't
+        // count system messages in ops count.
+        if (message.type === MessageType.NoOp || isSystemMessage(message)) {
             return;
         }
 
-        // We will queue a message to update our reference sequence number upon receiving a server
+        if (this.lastNoopTime === undefined) {
+            this.lastNoopTime = Date.now();
+        }
+
+        this.opsCountSinceNoop++;
+
+        // If the ops count since last op is greater than NoopCountFrequency and time since last noop is
+        // greater than NoopTimeFrequency, then send a Noop.
+        // We will send a message(Noop) to update our reference sequence number upon receiving a server
         // operation. This allows the server to know our true reference sequence number and be able to
         // correctly update the minimum sequence number (MSN).
-        if (this.updateSequenceNumberTimer === undefined) {
-            // Clear an update in 2 s
-            this.updateSequenceNumberTimer = setTimeout(() => {
-                this.updateSequenceNumberTimer = undefined;
-                if (this.activeConnection()) {
-                    this.submit(MessageType.NoOp, null);
-                }
-            }, CollabWindowTracker.NoopFrequency);
+        if (this.opsCountSinceNoop >= this.NoopCountFrequency
+            && this.activeConnection()
```

Comment:
Checking for this.activeConnection() is not required here - it's checked at the top of the function.

Reply:
Yeah.